apps/ejabberd/src/mongoose_cassandra_worker.erl

```diff
@@ -0,0 +1,480 @@
+%%%-------------------------------------------------------------------
+%%% @author Uvarov Michael <arcusfelis@gmail.com>
+%%% @copyright 2016 Erlang Solutions, Ltd.
+%%% @doc Generic cassandra worker
+%%% @end
+%%%-------------------------------------------------------------------
+-module(mongoose_cassandra_worker).
+
+%% ----------------------------------------------------------------------
+%% Exports
+
+%% API
+-export([cql_query/5,
+         cql_query_pool/5,
+         cql_query_async/5,
+         cql_query_pool_async/5,
+         cql_query_multi_async/5,
+         cql_query_pool_multi_async/5,
+         cql_batch/4,
+         cql_batch_pool/4,
+         cql_batch/5,
+         cql_batch_pool/5]).
+
+%% Helpers for debugging
+-export([test_query/1,
+         test_query/2,
+         queue_length/1,
+         queue_lengths/1,
+         total_count_query/2]).
+
+%% Internal exports
+-export([start_link/4]).
+
+%% mongoose_cassandra_worker callbacks
+-export([prepared_queries/0]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-behaviour(gen_server).
+-behaviour(mongoose_cassandra).
+
+%% ----------------------------------------------------------------------
+%% Imports
+
+-include_lib("ejabberd/include/ejabberd.hrl").
+-include_lib("ejabberd/include/jlib.hrl").
+-include_lib("exml/include/exml.hrl").
+
+-record(state, {
+    pool_name,
+    conn,
+    prepared_queries,
+    query_refs,
+    query_refs_count}).
+
+-record(prepared_query, {
+          query_id,
+          query_types}).
+
+%%====================================================================
+%% Internal functions
+%%====================================================================
+
+start_link(PoolName, Addr, Port, ClientOptions) ->
+    gen_server:start_link(?MODULE, [PoolName, Addr, Port, ClientOptions], []).
+
+%% @doc Do CQL query
+%%
+%% UserJID is used for rate-limiting, statistics and debugging
```

Comment:
Meaning of this JID is a partitioning key. Not a JID of user who makes the query. It can be used for some kind of locking, for example.
Currently it's only used in select_worker code. I.e. it's hashed to select worker.
I can add error reporting but backpressure and statistics is not that easy and require extending of the gen-server.
It also can be used to show warnings like this one:
```
issue=large_result_set, user=alice, number_of_elements=9000, query="SELECT ..."
```
The same information is inside query, but should be parsed. Here we can just use it.